pulsar-common/src/main/java/org/apache/pulsar/common/protocol/schema/SchemaHash.java

```diff
@@ -55,15 +82,123 @@ public static SchemaHash of(SchemaData schemaData) {
     }
 
     public static SchemaHash of(SchemaInfo schemaInfo) {
-        return of(schemaInfo == null ? new byte[0] : schemaInfo.getSchema(),
+        return of(schemaInfo == null ? null : schemaInfo.getSchema(),
                 schemaInfo == null ? null : schemaInfo.getType());
     }
 
-    public static SchemaHash of(byte[] schemaBytes, SchemaType schemaType) {
-        return new SchemaHash(hashFunction.hashBytes(schemaBytes), schemaType);
+    public static SchemaHash empty() {
+        return of(null, null);
+    }
+
+    private static SchemaHash of(byte[] schemaBytes, SchemaType schemaType) {
+        SchemaHash result;
+        if (schemaBytes == null || schemaBytes.length == 0) {
+            result = EmptySchemaHashFactory.get(schemaType);
+            if (result == null) {
+                log.warn("Could not get schemaHash from EmptySchemaHashFactory, will create by hashFunction. Might"
+                                + " bring performance regression. schemaBytes:{}, schemaType:{}",
+                        schemaBytes == null ? "null" : "0", schemaType);
+                result = new SchemaHash(
+                        hashFunction.hashBytes(schemaBytes == null ? new byte[0] : schemaBytes), schemaType);
+            }
+        } else {
+            result = new SchemaHash(hashFunction.hashBytes(schemaBytes), schemaType);
+        }
+        return result;
     }
 
     public byte[] asBytes() {
         return hash.asBytes();
     }
+
+    private static class EmptySchemaHashFactory {
```

Comment:
It may require some refactoring though why not having SchemaInfo.hash() method? The schemas are already all statically initialized

Reply:
+1, good idea.

Reply:
That sounds great, the code will be more clear @merlimat . I am trying add a SchemaInfo.hash() method to refactor  this patch.